src/shady/emit/c/emit_c.c

```diff
@@ -70,6 +70,36 @@ String legalize_c_identifier(Emitter* e, String src) {
     return string(e->arena, dst);
 }
 
+// For uniform constants, explicit locations require version 430 or ARB_explicit_uniform_location.
+static bool glsl_as_can_have_location(AddressSpace as) {
+    return as == AsUniformConstant || as == AsInput || as == AsUInput || as == AsOutput;
+}
+
+// Samplers and images (uniform constants)
+// Version 420 required.
+static bool glsl_as_can_have_binding(AddressSpace as) {
+    return as == AsUniformConstant || as == AsShaderStorageBufferObject || as == AsUniform;
+}
+
+static String glsl_try_make_single_int_qualifier(Emitter* emitter, const Node* decl, Nodes annotations, String annotation_name, String glsl_qualifier_name) {
+    const Node* a_node = lookup_annotation_list(annotations, annotation_name);
+    if (a_node) {
+        assert(is_annotation(a_node));
+        if (a_node->tag == AnnotationValue_TAG) {
+            const Node* value = a_node->payload.annotation_value.value;
+            if (value->tag == IntLiteral_TAG) {
```

Comment:
see also resolve_to_int_literal and get_int_literal_value for relevant helper fns